src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -163,7 +225,88 @@ private async Task GetProjectMetadataAsync(
             }
         }
 
-        private static void WarnForHttpSources(ListPackageArgs listPackageArgs, ListPackageProjectModel projectModel)
+        private static async Task<List<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>>> GetVulnerabilityData(
+            ListPackageProjectModel projectModel,
+            ListPackageReportModel reportModel,
+            IReadOnlyList<PackageSource> sources,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            var vulnerabilityInfo = new List<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>>();
+
+            foreach (var source in sources)
+            {
+                if (!await TryAddSourceVulnerabilityInfo(source, reportModel, logger, cancellationToken, vulnerabilityInfo))
+                {
+                    projectModel.AddProjectInformation(
+                        ProblemType.Warning,
+                        string.Format(CultureInfo.CurrentCulture, Strings.Warning_AuditSourceWithoutData, source.Name)
+                    );
+                }
+            }
+
+            return vulnerabilityInfo;
+        }
+
+        private static async Task<bool> TryAddSourceVulnerabilityInfo(
+            PackageSource source,
+            ListPackageReportModel reportModel,
+            ILogger logger,
+            CancellationToken cancellationToken,
+            List<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> vulnerabilityInfo)
+        {
+            var repository = Repository.Factory.GetCoreV3(source);
+            var vulnerabilityProvider = new VulnerabilityInfoResourceV3Provider();
+            var (isCreated, resource) = await vulnerabilityProvider.TryCreate(repository, cancellationToken);
+
+            if (!isCreated || resource is not VulnerabilityInfoResourceV3 vulnerabilityResource)
+            {
+                return false;
+            }
+
+            reportModel.SourcesUsed.Add(source);
+
+            var vulnerabilityInfoResult = await vulnerabilityResource.GetVulnerabilityInfoAsync(
+                new SourceCacheContext(),
+                logger,
+                cancellationToken
+            );
+
+            if (vulnerabilityInfoResult?.KnownVulnerabilities != null)
+            {
+                vulnerabilityInfo.AddRange(vulnerabilityInfoResult.KnownVulnerabilities);
+            }
+
+            return true;
+        }
+
+
+        private static IEnumerable<PackageVulnerabilityMetadata> GetPackageVulnerabilities(
```

Comment:
It doesn't look like any of this respects the NuGetAuditMode of `direct` vs `all`, or am I missing something?